source-snowflake/discovery.go

```diff
@@ -0,0 +1,383 @@
+package main
+
+import (
+	"context"
+	"database/sql"
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	pc "github.com/estuary/flow/go/protocols/capture"
+	pf "github.com/estuary/flow/go/protocols/flow"
+	"github.com/invopop/jsonschema"
+	"github.com/jmoiron/sqlx"
+	"github.com/sirupsen/logrus"
+	log "github.com/sirupsen/logrus"
+)
+
+func (snowflakeDriver) Discover(ctx context.Context, req *pc.Request_Discover) (*pc.Response_Discovered, error) {
+	var cfg = new(config)
+	if err := pf.UnmarshalStrict(req.ConfigJson, cfg); err != nil {
+		return nil, fmt.Errorf("error parsing config json: %w", err)
+	}
+	cfg.SetDefaults()
+
+	log.Info("performing discovery")
+
+	var db, err = connectSnowflake(ctx, cfg)
+	if err != nil {
+		return nil, fmt.Errorf("error connecting to snowflake: %w", err)
+	}
+	defer db.Close()
+
+	discoveryResults, err := performSnowflakeDiscovery(ctx, cfg, db)
+	if err != nil {
+		log.WithField("err", err).Fatal("error discovering snowflake collections")
+	}
+
+	bindings, err := catalogFromDiscovery(cfg, discoveryResults)
+	if err != nil {
+		log.WithField("err", err).Fatal("error converting discovery info to bindings")
+	}
+
+	return &pc.Response_Discovered{Bindings: bindings}, nil
+}
+
+type snowflakeDiscoveryResults struct {
+	Tables      []*snowflakeDiscoveryTable
+	Columns     []*snowflakeDiscoveryColumn
+	PrimaryKeys []*snowflakeDiscoveryPrimaryKey
+}
+
+type snowflakeDiscoveryTable struct {
+	Database string `db:"database_name"`
+	Schema   string `db:"schema_name"`
+	Name     string `db:"name"`
+	Kind     string `db:"kind"`
+	Comment  string `db:"comment"`
+}
+
+type snowflakeDiscoveryColumn struct {
+	Database         string  `db:"TABLE_CATALOG"`
+	Schema           string  `db:"TABLE_SCHEMA"`
+	Table            string  `db:"TABLE_NAME"`
+	Name             string  `db:"COLUMN_NAME"`
+	Index            int     `db:"ORDINAL_POSITION"`
+	DataType         string  `db:"DATA_TYPE"`
+	Comment          *string `db:"COMMENT"`
+	IsNullable       string  `db:"IS_NULLABLE"`
+	NumericScale     *int    `db:"NUMERIC_SCALE"`
+	NumericPrecision *int    `db:"NUMERIC_PRECISION"`
+}
+
+type snowflakeDiscoveryPrimaryKey struct {
+	Database   string `db:"database_name"`
+	Schema     string `db:"schema_name"`
+	Table      string `db:"table_name"`
+	ColumnName string `db:"column_name"`
+	KeySeq     int    `db:"key_sequence"`
+}
+
+func performSnowflakeDiscovery(ctx context.Context, cfg *config, db *sql.DB) (*snowflakeDiscoveryResults, error) {
+	log.Debug("performing discovery")
+
+	// We have to use `SHOW FOO` queries to get some of the information we need, such
+	// as primary keys. But there's no simple way to select just the columns we need
```

Comment:
There's [RESULT_SCAN](https://docs.snowflake.com/en/sql-reference/functions/result_scan), although maybe that is what you mean by not being a simple way, since it is a bit clunky.

Reply:
Yeah, that could probably be made to work but it seemed simpler to just use the `sqlx` package to automatically map result column names to struct fields, and I'd rather not change that now unless there's a compelling reason.

(The `sqlx` package was already an indirect dependency brought in by _something_ in the connectors repository, so trying to avoid another package dependency doesn't seem like enough reason IMO)